### PR TITLE
feat: structured logging (JSON)

### LINKED ISSUE #5107: Option to configure `UPTIME_KUMA_LOG_FORMAT`: `json` or `text`
### 📑 I have found these related issues/pull requests

- [x] Checked and didn't find similar issue

### 🏷️ Feature Request Type

Settings

### 🔖 Feature description

It would be very helpful for us to have an ENV_VAR to configure the logging format directly.
To be clear, a general option to switch between the current "text" logs and logs in "json" would already be a great benefit.

### ✔️ Solution

Extend `Logger` implementation with a e.g. `UPTIME_KUMA_LOG_FORMAT` variable that can be set to "text" or "json" with "text" being the default, similar to the implementation of `UPTIME_KUMA_HIDE_LOG`. 

Instead of logging to console as
```
console.error(timePart, modulePart, levelPart, msgPart);
```
with the value set to "json" the logging could log like this:
```
console.error(
  JSON.stringify(
    "timestamp": timePart,
    "module": modulePart,
    "level": levelPart,
    "message": msgPart
  )
);
```

### ❓ Alternatives

We have considered that this functionality would be possible by a log management solution, but a direct choice and configuration of the logFormat seems to us like a quick and valuable option to avoid further transformations in upstream solutions for the logs produced by kuma.

### 📝 Additional Context

n/a

### PR DESCRIPTION
**Pull request by SapphicCode**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Implements structured logging, as requested by #5107.
Resolves #5107

This feature can be enabled by using the environment variable `UPTIME_KUMA_LOG_FORMAT` set to `json`.

This PR also incorporates the work of #4391 and #4392 (with minor fixes to both as per the linter), for the feature itself I would direct your attention to commit 3e922be39cba2eba05ca814183c60def93e6ae2f.

Closes #4391
Closes #4392

## Type of change

Please delete any options that are not relevant.

- User interface (UI)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Before

Code - 2024-09-19 10 24 23

## After

Code - 2024-09-19 10 26 47

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

### CODE REVIEW

File: src/util.js [hunk 1]
```diff
@@ -155,98 +168,100 @@ class Logger {
...
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+            else {
+                timePart = now;
+                modulePart = `[${module}]`;
+                levelPart = `${level}:`;
```
Comment (CommanderStorm) [inline src/util.js:216]: This is alredy the value, right?

```suggestion
                modulePart = `[${module}]`;
```

**Review by CommanderStorm — COMMENTED** [review]
Hi, overall, this looks excellent.
I have left two minor style things below

File: src/util.js (hunk 1 above)
Comment (SapphicCode) [reply src/util.js:216]: You're right, the `timePart` is redundant. The other two are actually modifying things though.

Comment (CommanderStorm) [reply src/util.js]: Good Point, overlooked the second one

**Review by CommanderStorm — APPROVED** [review]
Thanks for the refactoring/logging improvement! 🎉 



> [!NOTE]
> This PR is part of the [`v2.0` merge window](https://github.com/louislam/uptime-kuma/milestone/24) => see #4500 for the bugs that need to be addressed before we can ship this release ^^

<sub>All changes in this PR are small and uncontroversial ⇒ merging with junior maintainer approval</sub>

File: src/util.ts [hunk 2]
```diff
@@ -233,10 +222,23 @@ class Logger {
...
-        let timePart: string;
-        let modulePart: string;
-        let levelPart: string;
-        let msgPart: string;
+        let timePart: string = now;
+        let modulePart: string = module;
+        let levelPart: string = level;
+        let msgPart: unknown = msg;
+
+        if (process.env.UPTIME_KUMA_LOG_FORMAT === "json") {
```
Comment (louislam) [inline src/util.ts:230]: Since this function is not backend only, it causes issue like #5157.

Comment (louislam) [reply src/util.ts:230]: It has been fixed by adding `process.env` as an empty object in vite.config.js.

**Comment by louislam** [comment]
I think it also breaks the hide logging feature.

File: server/model/monitor.js [hunk 3]
```diff
@@ -1379,11 +1380,12 @@ class Monitor extends BeanModel {
...
             }
 
-            let notifyDays = await setting("tlsExpiryNotifyDays");
+            let notifyDays = await Settings.get("tlsExpiryNotifyDays");
             if (notifyDays == null || !Array.isArray(notifyDays)) {
                 // Reset Default
-                await setSetting("tlsExpiryNotifyDays", [ 7, 14, 21 ], "general");
+                await Settings.set("tlsExpiryNotifyDays", [ 7, 14, 21 ], "general");
                 notifyDays = [ 7, 14, 21 ];
+                await Settings.set("tlsExpiryNotifyDays", notifyDays, "general");
```
Comment (louislam) [inline server/model/monitor.js:1388]: This line is not presented in the old code.

Comment (SapphicCode) [reply server/model/monitor.js:1388]: I suppose that would be a question for @CommanderStorm in #4391: https://github.com/louislam/uptime-kuma/pull/4391/files#diff-0c4874f149ad01732ad730f36c95c99e37f5c3f13e9b52f628e2e284daa72d27R1388

Comment (CommanderStorm) [reply server/model/monitor.js:1388]: image

Yea, the first time of setting the setting should not exist.

**Comment by SapphicCode** [comment]
> I think it also breaks the hide logging feature.

I'm not sure how that's possible, those lines weren't modified at all...

File: server/model/monitor.js (hunk 3 above)
Comment (SapphicCode) [reply server/model/monitor.js:1388]: Fixed: https://github.com/SapphicCode/fork-uptime-kuma/commit/861e100d82a8ed4aaa860259dae0f6d97549213a

**Comment by CommanderStorm** [comment]
+1 this does not break the hide log feature.

https://github.com/louislam/uptime-kuma/issues/5157 is a pretty unfortunate thing, but I don't think the whole console logging filtering code should be executed in the browser.
We should remove that part (in a different PR).

I think pushing this is not nessesary, but don't want to fight about this
=> lets push this for one release.

**Pull request merged**